projects/gloo/pkg/bootstrap/clients/config_factory.go

```diff
@@ -0,0 +1,116 @@
+package clients
+
+import (
+	"path/filepath"
+
+	consulapi "github.com/hashicorp/consul/api"
+	errors "github.com/rotisserie/eris"
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/k8s-utils/kubeutils"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/factory"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/kube"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/kube/crd"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/memory"
+	"k8s.io/client-go/rest"
+)
+
+type (
+	configFactoryParamsMemory struct {
+		sharedCache memory.InMemoryResourceCache
+	}
+	configFactoryParamsKube struct {
+		kubeCache kube.SharedCache
+		restCfg   **rest.Config
+	}
+	configFactoryParamsConsul struct {
+		consulClient *consulapi.Client
+	}
+
+	ConfigFactoryParams struct {
+		settings *v1.Settings
+		memory   configFactoryParamsMemory
+		kube     configFactoryParamsKube
+		consul   configFactoryParamsConsul
+	}
+)
+
+func NewConfigFactoryParams(settings *v1.Settings,
+	sharedCache memory.InMemoryResourceCache,
+	cache kube.SharedCache,
+	cfg **rest.Config,
+	consulClient *consulapi.Client) ConfigFactoryParams {
+	return ConfigFactoryParams{
+		settings: settings,
+		memory: configFactoryParamsMemory{
+			sharedCache: sharedCache,
+		},
+		kube: configFactoryParamsKube{
+			kubeCache: cache,
+			restCfg:   cfg,
+		},
+		consul: configFactoryParamsConsul{
+			consulClient: consulClient,
+		},
+	}
+}
+
+// sharedCache, resourceCrd+cfg OR consulClient must be non-nil
```

Comment:
This elucidates the required non-nils. I think it's

```suggestion
// params.memory.sharedCache, resourceCrd+params.kube.restCfg OR params.consul.consulClient must be non-nil
```